### PR TITLE
FEAT: Add new spec thp_shmem_enabled

### PR DESCRIPTION
**Pull request by TZ3070**
Signed-off-by: Xinting Li <xintli@redhat.com>

rh-pre-commit.version: 2.3.2
rh-pre-commit.check-secrets: ENABLED

### All Pull Requests:

Check all that apply:

* [x] Have you followed the guidelines in our Contributing document, including the instructions about commit messages?
* [x] No Sensitive Data in this change?
* [ ] Is this PR to correct an issue?
* [ ] Is this PR an enhancement?

### Complete Description of Additions/Changes:

*Add your description here*
RHINENG-24776

Add new spec thp_shmem_enabled and new parser ThpShmemEnabled

## Summary by Sourcery

Add support for parsing the transparent hugepage shmem_enabled setting and expose it as a new spec.

New Features:
- Introduce the ThpShmemEnabled parser to read the transparent hugepage shmem_enabled configuration.
- Add a new thp_shmem_enabled spec pointing to /sys/kernel/mm/transparent_hugepage/shmem_enabled.

Tests:
- Extend transparent hugepage parser tests to cover the new ThpShmemEnabled parser.

## Summary by Sourcery

Add support for collecting and parsing the transparent hugepage shmem_enabled setting as a new spec and parser, and expand documentation and tests for transparent hugepage parsers.

New Features:
- Add a ThpShmemEnabled parser to expose the active transparent hugepage shmem_enabled option.
- Introduce a thp_shmem_enabled spec mapped to /sys/kernel/mm/transparent_hugepage/shmem_enabled.

Enhancements:
- Improve transparent hugepage parser documentation and doctest coverage for existing parsers.

Tests:
- Extend transparent hugepage parser tests to cover ThpShmemEnabled, including empty and malformed input cases.
- Add doctest-based tests to validate documented examples for transparent hugepage parsers.

### CODE REVIEW

File: insights/parsers/transparent_hugepage.py [hunk 1]
```diff
@@ -73,3 +80,34 @@ def parse_content(self, content):
         for w in self.line.split():
             if w.startswith("[") and w.endswith("]"):
                 self.active_option = w[1:-1]
+
+
+@parser(Specs.thp_shmem_enabled)
+class ThpShmemEnabled(Parser):
+    """
+    Class for parsing the ``/sys/block/*/queue/max_segment_size`` files..
```
Comment (JoySnow) [inline insights/parsers/transparent_hugepage.py]: seems a typo, guess this mean to be the new spec path ...

File: insights/parsers/transparent_hugepage.py [hunk 2]
```diff
@@ -33,14 +35,19 @@
...
-    >>> shared[ThpEnabled].active_option
-    madvise
+    >>> type(thp_enabled)
+    <class 'insights.parsers.transparent_hugepage.ThpEnabled'>
+    >>> thp_enabled.line
+    'always [madvise] never'
+    >>> thp_enabled.active_option
+    'madvise'
 
+ThpEnabled - file ``/sys/kernel/mm/transparent_hugepage/shmem_enabled``
```
Comment (JoySnow) [inline insights/parsers/transparent_hugepage.py]: seems a typo here, you may want to change `ThpEnabled` to `ThpShmemEnabled`

File: insights/parsers/transparent_hugepage.py [hunk 3]
```diff
@@ -73,3 +80,34 @@ def parse_content(self, content):
...
+        'always within_size advise [never] deny force'
+        >>> thp_shmem_enabled.active_option
+        'never'
+    """
+    def parse_content(self, content):
+        self.line = " ".join(content).strip()
+        self.active_option = None
+        for w in self.line.split():
+            if w.startswith("[") and w.endswith("]"):
+                self.active_option = w[1:-1]
```
Comment (JoySnow) [inline insights/parsers/transparent_hugepage.py:102]: shall we give it a `break` here to stop the for loop on the first found of `[xxx]` value?

File: insights/tests/parsers/test_transparent_hugepage.py [hunk 4]
```diff
@@ -27,6 +32,10 @@
 # testing without newlines
 ENABLED_ALWAYS = """[always] madvise never"""
 
+THP_SHMEM_ENABLED_ALWAYS = """[always] within_size advise never deny force"""
+THP_SHMEM_ENABLED_NEVER = """always within_size advise [never] deny force"""
```
Comment (JoySnow) [inline insights/tests/parsers/test_transparent_hugepage.py:36]: Would it make sense if adding some corner cases like:
- `"abc def xyz"`
- `"abc def [] xyz"`
- `"abc def [ ] xyz"`

**Review by JoySnow — CHANGES_REQUESTED** [review]
hey @TZ3070 , pls check my comments inline

File: insights/tests/parsers/test_transparent_hugepage.py (hunk 4 above)
Comment (TZ3070) [reply insights/tests/parsers/test_transparent_hugepage.py:36]: Updated

File: insights/parsers/transparent_hugepage.py (hunk 2 above)
Comment (TZ3070) [reply insights/parsers/transparent_hugepage.py]: updated

File: insights/parsers/transparent_hugepage.py (hunk 3 above)
Comment (TZ3070) [reply insights/parsers/transparent_hugepage.py:102]: updated

File: insights/parsers/transparent_hugepage.py (hunk 1 above)
Comment (TZ3070) [reply insights/parsers/transparent_hugepage.py]: updated

**Review by JoySnow — APPROVED** [review]

File: insights/parsers/transparent_hugepage.py [hunk 5]
```diff
@@ -33,14 +35,19 @@
...
-    >>> shared[ThpEnabled].line
-    always [madvise] never
-    >>> shared[ThpEnabled].active_option
-    madvise
+    >>> type(thp_enabled)
+    <class 'insights.parsers.transparent_hugepage.ThpEnabled'>
+    >>> thp_enabled.line
+    'always [madvise] never'
+    >>> thp_enabled.active_option
+    'madvise'
```
Comment (xiangce) [inline insights/parsers/transparent_hugepage.py]: Since we have to update this docstring's change for testing, would you mind moving these docstring of existing Parsers to the Parser's definition below?

Comment (TZ3070) [reply insights/parsers/transparent_hugepage.py]: updated

**Review by xiangce — APPROVED** [review]

**Comment by xiangce** [comment]
test dp

**Comment by xiangce** [comment]
test egg

**Comment by xiangce** [comment]
test rpm

**Comment by qinpingli** [comment]
Close and reopen this PR to trigger the checks.

**Review by qinpingli — APPROVED** [review]
Per RHINENG-24776, all tests are passed. Approved this PR.

**Comment by qinpingli** [comment]
Created a new PR same with this: https://github.com/RedHatInsights/insights-core/pull/4785.

I'll close this PR.
